lib/memo_wise.rb

```diff
@@ -114,6 +114,163 @@ def self.has_kwarg?(method) # rubocop:disable Naming/PredicateName
     end
   end
 
+  # @private
+  #
+  # Returns scope, visibility, class, and method object for a given method name.
+  #
+  # @param klass [Class]
+  #   Class in which to search for a method by name.
+  #
+  # @param method_name [Symbol]
+  #   Name of method to search for at both class and instance levels.
+  #
+  # @return [(Symbol, Symbol, Class, UnboundMethod)]
+  #   Return **scope**, **visibility**, **class**, and **method** corresponding
+  #   to the given method name on the target class.
+  #
+  #     - **scope**      -- either `:class` or `:instance`
+  #     - **visibility** -- either `:private`, `:protected`, or `:public`
+  #     - **class**      -- either `klass` or its
+  #                         [singleton class](https://medium.com/@leo_hetsch/demystifying-singleton-classes-in-ruby-caf3fa4c9d91)
+  #     - **method**     -- instance of {UnboundMethod}
+  #
+  # @raise ArgumentError
+  #   Raises `ArgumentError` unless `method_name` is a `Symbol` corresponding
+  #   to a class or instance method defined on `klass`.
+  #
+  # rubocop:disable Layout/LineLength
+  # @example
+  #   class Example
+  #     def an_instance_method; end
+  #
+  #     class << self
+  #       def a_class_method; end
+  #       private :a_class_method
+  #     end
+  #   end
+  #
+  #   MemoWise.method_info(Example, :an_instance_method)[0, 2] #=> [:instance, :public]
+  #
+  #   MemoWise.method_info(Example, :a_class_method)[0, 2] #=> [:class, :private]
+  # rubocop:enable Layout/LineLength
+  #
+  def self.method_info(klass, method_name)
+    unless klass.is_a?(Class) || klass.is_a?(Module)
+      raise ArgumentError, "#{klass.inspect} must be a Class or Module"
+    end
+
+    unless method_name.is_a?(Symbol)
+      raise ArgumentError, "#{method_name.inspect} must be a Symbol"
+    end
+
+    # In Ruby, "class methods" are implemented as normal instance methods
+    # on the "singleton class" of a given Class object, found via
+    # {Class#singleton_class}.
+    # See: https://medium.com/@leo_hetsch/demystifying-singleton-classes-in-ruby-caf3fa4c9d91
+    singleton = klass.singleton_class
+
+    is_singleton_method =
+      singleton.private_instance_methods.include?(method_name) ||
+      singleton.instance_methods.include?(method_name)
+
+    if is_singleton_method
```

Comment:
Would we want to consider making this the `else` case so that when the same name exists for both a class and instance method we default to memoizing the instance one (which I think will be more common)?